src/main/java/org/asteriskjava/manager/internal/ManagerConnectionImpl.java

```diff
@@ -1534,6 +1535,48 @@ private void reconnect()
         }
     }
 
+    private void cleanupActionListeners(DisconnectEvent event)
+    {
+        ArrayList<String> discardedInternalActionIdList = new ArrayList<String>();
+        synchronized (responseListeners)
+        {
+            // Clear response listeners that will not receive their responses
+            for (Map.Entry<String,  SendActionCallback> entry : responseListeners.entrySet())
+            {
+                discardedInternalActionIdList.add(entry.getKey());
+                SendActionCallback responseListener = entry.getValue();
+                // Allows to unblock waiting sendAction() calls
+                responseListener.onResponse(null);
+            }
+            responseListeners.clear();
+        }
+
+        synchronized (responseEventListeners)
+        {
+            // Remove those already cleaned up via responseListeners
+            for (String discardedInternalActionId: discardedInternalActionIdList)
+            {
+                responseEventListeners.remove(discardedInternalActionId);
+            }
+            // Remove remaining. These could be EventGeneratingAction
+            // handlers that have received a response but have not yet
+            // received the end event.
+            for (ManagerEventListener responseEventListener: responseEventListeners.values())
+            {
+                try
+                {
+                    // Allows to unblock waiting sendAction() calls
+                    responseEventListener.onManagerEvent(event);
```

Comment:
This is concerning and could lead to a dead lock, depending on what the call to responseEventListener.onManagerEvent() call does.

It would be better to add the responseEventListeners to another list while holding the lock, and then call each of the listeners after releasing the lock something like this (forgive my short hand)...

toCallback = new List();
synchronized(listeners)
{
   for (...)
      toCallback.add(listener);
}
for (toCallback)
    toCallback.onManagerEvent(..)

Reply:
Thank you for looking into it. I can't look into it closely right now but your comment makes sense. I aim to fix that in a day or two.

Reply:
I'll be making a new pull request to save on diff noise. This pull request should be closed without merging.